### PR TITLE
Codefix for removing Unused Identifiers 

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Quick fix for removing unused variables and unused locals.

Current known limitations:
* no support for the spread operator
* no support for the `for ... in` loop

### CODE REVIEW

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                        return createCodeFix("{}", initializer.declarations[0].pos, initializer.declarations[0].end - initializer.declarations[0].pos);
+                    }
+                    else if (token.parent.parent.kind === SyntaxKind.CatchClause) {
+                        const catchClause = <CatchClause>token.parent.parent;
+                        const parameter = catchClause.variableDeclaration.getChildren()[0];
+                        return createCodeFix("", parameter.pos, parameter.end - parameter.pos);
+                    }
+                    else {
+                        const variableStatement = <VariableStatement>token.parent.parent.parent;
+                        if (variableStatement.declarationList.declarations.length === 1) {
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: consider sharing the logic here with the other list removal cases.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            const sourceFile = context.sourceFile;
+            const start = context.span.start;
+            const token = getTokenAtPosition(sourceFile, start);
+
+            if (token.kind === ts.SyntaxKind.Identifier) {
+                if (token.parent.kind === ts.SyntaxKind.VariableDeclaration) {
+                    if (token.parent.parent.parent.kind === SyntaxKind.ForStatement) {
+                        const forStatement = <ForStatement>token.parent.parent.parent;
+                        const initializer = <VariableDeclarationList>forStatement.initializer;
+                        if (initializer.declarations.length === 1) {
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: consider merging this into a list removal logic for all of the cases below.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                            const declarations = variableStatement.declarationList.declarations;
+                            if (declarations[0].name === token) {
+                                return createCodeFix("", token.parent.pos + 1, token.parent.end - token.parent.pos);
+                            }
+                            else {
+                                return createCodeFix("", token.parent.pos - 1, token.parent.end - token.parent.pos + 1);
+                            }
+                        }
+                    }
+                }
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: what about binding patterns? can you add tests for these cases. both for parameters and variable declarations.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+                if (token.parent.kind === SyntaxKind.FunctionDeclaration ||
+                    token.parent.kind === SyntaxKind.ClassDeclaration ||
+                    token.parent.kind === SyntaxKind.InterfaceDeclaration ||
+                    token.parent.kind === SyntaxKind.MethodDeclaration ||
+                    token.parent.kind === SyntaxKind.ModuleDeclaration ||
+                    token.parent.kind === SyntaxKind.PropertyDeclaration ||
+                    token.parent.kind === SyntaxKind.ArrowFunction) {
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: Also please add handeling for computed propoerties with non-identifier names:

```ts
class C {
    private ["string"] (){}
    private [Symbol.Iterator]() {}
}
```

File: src/services/codefixes/unusedIdentifierFixes.ts (hunk 1 above)
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: done

File: src/services/codefixes/unusedIdentifierFixes.ts (hunk 2 above)
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: doen

File: tests/cases/fourslash/unusedParameterInConstructor1.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+/// <reference path='fourslash.ts' />
+
+// @noUnusedLocals: true
+//// class C1 {
+////    [|constructor(private p1: string, public p2: boolean, public p3: any, p5)|] { p5; }
+//// }
+
+verify.codeFixAtPosition("constructor(public p2: boolean, public p3: any, p5)");
```
Comment (paulvanbrenk) [inline tests/cases/fourslash/unusedParameterInConstructor1.ts]: added

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                            return createCodeFix("", token.parent.pos, token.parent.end - token.parent.pos + 1);
+                        }
+                        else {
+                            return createCodeFix("", token.parent.pos - 1, token.parent.end - token.parent.pos + 1);
+                        }
+                    }
+                }
+
+                if (token.parent.kind === ts.SyntaxKind.Parameter) {
+                    const functionDeclaration = <FunctionDeclaration>token.parent.parent;
```
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: done

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                    token.parent.kind === SyntaxKind.InterfaceDeclaration ||
+                    token.parent.kind === SyntaxKind.MethodDeclaration ||
+                    token.parent.kind === SyntaxKind.ModuleDeclaration ||
+                    token.parent.kind === SyntaxKind.PropertyDeclaration ||
+                    token.parent.kind === SyntaxKind.ArrowFunction) {
+                    return createCodeFix("", token.parent.pos, token.parent.end - token.parent.pos);
+                }
+
+                if (token.parent.kind === SyntaxKind.TypeParameter) {
+                    const typeParameters = (<ClassDeclaration>token.parent.parent).typeParameters;
```
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: Used DeclWIthTypeParam, tests were already there.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+                            }
+                            else {
+                                return createCodeFix("", token.parent.pos - 1, token.parent.end - token.parent.pos + 1);
+                            }
+                        }
+                    }
+                    else if (token.parent.parent.parent.kind === SyntaxKind.ForInStatement) {
+                        const forInStatement = <ForInStatement>token.parent.parent.parent;
+                        const initializer = <VariableDeclarationList>forInStatement.initializer;
+                        return createCodeFix("{}", initializer.declarations[0].pos, initializer.declarations[0].end - initializer.declarations[0].pos);
```
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: Doing nothing, modified test to make sure.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                            return createCodeFix("", token.parent.pos - 1, token.parent.end - token.parent.pos + 1);
+                        }
+                    }
+                }
+
+                if (token.parent.parent.kind === SyntaxKind.ImportClause || token.parent.parent.kind === SyntaxKind.ImportDeclaration) {
+                    return createCodeFix("{}", token.parent.pos, token.parent.end - token.parent.pos);
+                }
+
+                if (token.parent.kind === SyntaxKind.ImportEqualsDeclaration) {
```
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: This is `import a = A`

File: tests/cases/fourslash/unusedLocalsInFunction2.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+/// <reference path='fourslash.ts' />
+
+// @noUnusedLocals: true
+////function greeter() {
+////    [| var x, y = 0; |]
+////    x++;
+////}
+
+verify.codeFixAtPosition("var x;", 6133);
```
Comment (RyanCavanaugh) [inline tests/cases/fourslash/unusedLocalsInFunction2.ts]: No error codes in fourslash tests

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                                case SyntaxKind.ForOfStatement:
+                                    const forOfStatement = <ForOfStatement>token.parent.parent.parent;
+                                    if (forOfStatement.initializer.kind === SyntaxKind.VariableDeclarationList) {
+                                        const forOfInitializer = <VariableDeclarationList>forOfStatement.initializer;
+                                        return createCodeFix("{}", forOfInitializer.declarations[0].pos, forOfInitializer.declarations[0].end - forOfInitializer.declarations[0].pos);
+                                    }
+                                    break;
+
+                                case SyntaxKind.ForInStatement:
+                                    // There is no valid fix in the case of:
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: rename the variable to start with `_`.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                                    if (variableStatement.declarationList.declarations.length === 1) {
+                                        return createCodeFix("", variableStatement.pos, variableStatement.end - variableStatement.pos);
+                                    }
+                                    else {
+                                        const declarations = variableStatement.declarationList.declarations;
+                                        return removeSingleItem(declarations, token);
+                                    }
+                            }
+
+                        case SyntaxKind.FunctionDeclaration:
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: namespace, type alais, and enum as well

Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: i would move this to the default clause and use `isDeclarationName`

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                                return createCodeFix("", token.parent.pos, token.parent.end - token.parent.pos);
+                            }
+                            else {
+                                return removeSingleItem(functionDeclaration.parameters, token);
+                            }
+
+                        case SyntaxKind.ImportSpecifier:
+                            const namedImports = <NamedImports>token.parent.parent;
+                            const elements = namedImports.elements;
+                            if (elements.length === 1) {
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: if no other imports in this list we need to remove the whole module import.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+                        // handle case where 'import * as a from './file'
+                        case SyntaxKind.NamespaceImport:
+                            return createCodeFix("", token.parent.parent.parent.pos, token.parent.parent.parent.end - token.parent.parent.parent.pos);
+
+                        case SyntaxKind.EnumDeclaration:
+                            return createCodeFix("", token.parent.pos, token.parent.end - token.parent.pos);
+                    }
+                    break;
+
+                case SyntaxKind.PrivateKeyword:
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: When is this ever used?

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+            Diagnostics._0_is_declared_but_never_used.code,
+            Diagnostics.Property_0_is_declared_but_never_used.code
+        ],
+        getCodeActions: (context: CodeFixContext) => {
+            const sourceFile = context.sourceFile;
+            const start = context.span.start;
+            const token = getTokenAtPosition(sourceFile, start);
+
+            switch (token.kind) {
+                case ts.SyntaxKind.Identifier:
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: binding patterns are not handled anywhere:

```ts
function f({a, b}) {}
```

Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: Also please add handeling for computed properties with non-identifier names:

```ts
class C {
    private ["string"] (){}
    private [Symbol.Iterator]() {}
}
```

File: tests/cases/fourslash/unusedImports12FS.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+/// <reference path='fourslash.ts' />
+
+// @noUnusedLocals: true
+// @Filename: file2.ts
+//// [| import f1, * as s from "./file1"; |]
```
Comment (mhegazy) [inline tests/cases/fourslash/unusedImports12FS.ts]: Please add a test with both unused.

File: src/services/codefixes/unusedIdentifierFixes.ts [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+                        case SyntaxKind.EnumDeclaration:
+                            return createCodeFix("", token.parent.pos, token.parent.end - token.parent.pos);
+                    }
+                    break;
+
+                case SyntaxKind.PrivateKeyword:
+                case SyntaxKind.PropertyDeclaration:
+                    return createCodeFix("", token.parent.pos, token.parent.end - token.parent.pos);
+
+                case SyntaxKind.AsteriskToken:
```
Comment (mhegazy) [inline src/services/codefixes/unusedIdentifierFixes.ts]: when do we get a `*` token?

File: src/services/codefixes/unusedIdentifierFixes.ts (hunk 11 above)
Comment (paulvanbrenk) [inline src/services/codefixes/unusedIdentifierFixes.ts]: Out of scope, will add for 2nd iteration of this fix

File: tests/cases/fourslash/unusedImports12FS.ts (hunk 16 above)
Comment (paulvanbrenk) [inline tests/cases/fourslash/unusedImports12FS.ts]: That generates 2 errors and 2 fixes. So not applicable

**Pull request merged**